### PR TITLE
#659 stack slots count uses actual field size

### PR DESCRIPTION
**Pull request by jarekratajski**
this fixes my problem

### CODE REVIEW

File: compiler/ETA/DeSugar/DsForeign.hs [hunk 1]
```diff
@@ -655,19 +655,20 @@ dsFExport closureId inheritsFamTyCon famInstEnvs co externalName classSpec mod =
...
-                              <> invokespecial
-                                  (mkMethodRef argClass "<init>" [argPrimFt] void)))
-                    mempty
-                    (zip3 [localVariableStart..] argFts argTypes)
+        else fst $ foldl' (\(code, stackIndex) (i, argPrimFt, argType) ->
+                           let argClass = typeDataConClass dflags extendsInfo argType
+                               argClassFt = obj argClass
+                           in (code
+                           <> (if argPrimFt == jbool
+                               then gload argPrimFt i <> ifeq falseClosure trueClosure
```
Comment (rahulmutt) [inline compiler/ETA/DeSugar/DsForeign.hs]: With the changes you've made, this `i` should probably be replaced with `stackIndex`.

File: compiler/ETA/DeSugar/DsForeign.hs [hunk 2]
```diff
@@ -655,19 +655,20 @@ dsFExport closureId inheritsFamTyCon famInstEnvs co externalName classSpec mod =
...
+                               else new argClassFt
+                                 <> dup argClassFt
+                                 <> gload argPrimFt stackIndex
+                                 <> invokespecial
+                                     (mkMethodRef argClass "<init>" [argPrimFt] void))
+                               , stackIndex + fieldSize argPrimFt))
+                           (mempty, localVariableStart)
+                           (zip3 [localVariableStart..] argFts argTypes)
```
Comment (rahulmutt) [inline compiler/ETA/DeSugar/DsForeign.hs]: The `[localVariableStart..]` (`i`) is not even used anymore we can remove it to reduce the entire expression from `zip3` to `zip`.

Comment (jarekratajski) [reply compiler/ETA/DeSugar/DsForeign.hs]: Thanks. Sure - have not seen the problem. I am now testing this fix on my projects.

**Comment by rahulmutt** [comment]
There were some build failures, but that is not related to this commit. Before I can merge this commit, I'll need you to digitally sign the CLA. You can find a link in the 2nd comment of this pull request.

**Comment by jarekratajski** [comment]
signed. And works for me.

**Pull request merged**